// request-handler/request-handler.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/http'), require('ng2-interceptors/index'), require('angular2-jwt'), require('@angular/http/testing')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/http', 'ng2-interceptors/index', 'angular2-jwt', '@angular/http/testing'], factory) :
	(factory((global['request-handler'] = {}),global.core,global.common,global.http,global.index,global.angular2Jwt,global.testing));
}(this, (function (exports,core,common,http,index,angular2Jwt,testing) { 'use strict';

var HANDLER_CONFIG = new core.OpaqueToken('handler.config');

/*
 * Third party
 */
/**
 * Models
 */
var RequestHandlerService = (function () {
    /**
     * @param {?} injector
     * @param {?} handlerConfig
     */
    function RequestHandlerService(injector, handlerConfig) {
        this.injector = injector;
        this.handlerConfig = handlerConfig;
        this.jwtHelper = new angular2Jwt.JwtHelper();
    }
    Object.defineProperty(RequestHandlerService.prototype, "http", {
        /**
         * Injects http.
         * @return {?}
         */
        get: function () {
            return this.injector.get(http.Http);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RequestHandlerService.prototype, "config", {
        /**
         * Config helper function for testing purposes.
         * @return {?}
         */
        get: function () {
            return this.handlerConfig;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Run before any http request is made
     *
     * You can return:
     *   - Request: The modified request
     *   - Nothing: For convenience: It's just like returning the request
     *   - <any>(Observable.throw("cancelled")): Cancels the request,
     *     interrupting it from the pipeline, and calling back 'interceptAfter'
     *     in backwards order of those interceptors that got called up to this point.
     * @param {?} request
     * @return {?}
     */
    RequestHandlerService.prototype.interceptBefore = function (request) {
        var _this = this;
        /**
         * Check the request for the presence of a token:
         *  - If no token is present, continue the call as per usual.
         */
        var accessToken = request.options.headers.get(this.config.clientAccessTokenName);
        if (!accessToken || !this.jwtHelper.isTokenExpired(accessToken)) {
            return request;
        }
        var /** @type {?} */ currentCredentials = JSON.parse(localStorage.getItem('auth.headers'));
        /**
         * If there is no access token in local storage,
         * Reassign currentCredentials those in session storage.
         */
        if (!currentCredentials || (currentCredentials && currentCredentials[this.handlerConfig.clientRefreshTokenName] == null)) {
            currentCredentials = JSON.parse(sessionStorage.getItem('auth.headers'));
        }
        /**
         * Set the headers for the request.
         * If the refresh token is not present, throw an error.
         */
        var headers = new http.Headers();
        if (currentCredentials && currentCredentials[this.handlerConfig.clientRefreshTokenName]) {
            headers.append(this.handlerConfig.serverRefreshTokenName, currentCredentials[this.handlerConfig.clientRefreshTokenName]);
        }
        else {
            throw Error('No Refresh Token available');
        }
        /**
         *  Calls the refresh endpoint
         */
        var obs = this.http.post(this.handlerConfig.refreshUrl, {}, {
            headers: headers,
        }).map(function (res) {
            /**
             *  Handles the response from the endpoint
             *  Returns the original request, which follows the refresh pipeline
             */
            var localItem = JSON.parse(localStorage.getItem('auth.headers'));
            var persist = localItem !== null && localItem[_this.handlerConfig.clientAccessTokenName] !== null;
            // Populate the Auth Headers Object
            var authHeaders = {};
            authHeaders[_this.handlerConfig.clientAccessTokenName] = res.headers.get(_this.handlerConfig.serverAccessTokenName);
            authHeaders[_this.handlerConfig.clientRefreshTokenName] = res.headers.get(_this.handlerConfig.serverRefreshTokenName);
            /**
             * Save to persisiten storage if the current auth items are stored there.
             */
            if (persist) {
                localStorage.setItem('auth.headers', JSON.stringify(authHeaders));
                sessionStorage.removeItem('auth.headers');
            }
            else {
                sessionStorage.setItem('auth.headers', JSON.stringify(authHeaders));
                localStorage.removeItem('auth.headers');
            }
            request.options.headers.set(_this.handlerConfig.serverAccessTokenName, authHeaders[_this.handlerConfig.clientAccessTokenName]);
            return request;
        });
        return obs;
    };
    /**
     * @param {?} response
     * @return {?}
     */
    RequestHandlerService.prototype.interceptAfter = function (response) {
        return response;
    };
    return RequestHandlerService;
}());
RequestHandlerService.decorators = [
    { type: core.Injectable },
];
/**
 * @nocollapse
 */
RequestHandlerService.ctorParameters = function () { return [
    { type: core.Injector, },
    { type: undefined, decorators: [{ type: core.Inject, args: [HANDLER_CONFIG,] },] },
]; };

var validToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE1MDc3Mjc3NTcsImV4cCI6MTUwNzczMjUyMywiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSIsIkdpdmVuTmFtZSI6IkpvaG5ueSIsIlN1cm5hbWUiOiJSb2NrZXQiLCJFbWFpbCI6Impyb2NrZXRAZXhhbXBsZS5jb20iLCJSb2xlIjpbIk1hbmFnZXIiLCJQcm9qZWN0IEFkbWluaXN0cmF0b3IiXX0.XSlAo8SL-pLBK4eLCPLAGZ4Mz732x2yGAM4WW_PwmYg';

/**
 * @param {?} backend
 * @param {?} options
 * @param {?} realBackend
 * @return {?}
 */
function fakeBackendFactory(backend, options, realBackend) {
    backend.connections.subscribe(function (connection) {
        /**
         * Simulate response time
         */
        if (connection.request.url.substring(0, 7) === 'test://' && connection.request.method === http.RequestMethod.Post) {
            var /** @type {?} */ headers = new http.Headers({
                'X-Access-Token': validToken,
            });
            connection.mockRespond(new http.Response(new http.ResponseOptions({
                status: 200,
                headers: headers,
            })));
        }
        else {
            var /** @type {?} */ realHttp = new http.Http(realBackend, options);
            var /** @type {?} */ requestOptions = new http.RequestOptions({
                method: connection.request.method,
                headers: connection.request.headers,
                body: connection.request.getBody(),
                url: connection.request.url,
                withCredentials: connection.request.withCredentials,
                responseType: connection.request.responseType
            });
            realHttp.request(connection.request.url, requestOptions)
                .subscribe(function (response) {
                connection.mockRespond(response);
            }, function (error) {
                connection.mockError(error);
            });
        }
        
        return;
    });
    return new http.Http(backend, options);
}

var fakeBackendProvider = {
    // use fake backend in place of Http service for backend-less development
    provide: http.Http,
    useFactory: fakeBackendFactory,
    deps: [testing.MockBackend, http.BaseRequestOptions, http.XHRBackend]
};

/**
 * @param {?} xhrBackend
 * @param {?} requestOptions
 * @param {?} requestHandlerService
 * @return {?}
 */
function InterceptorFactory(xhrBackend, requestOptions, requestHandlerService) {
    var /** @type {?} */ service = new index.InterceptorService(xhrBackend, requestOptions);
    service.addInterceptor(requestHandlerService);
    return service;
}
var RequestHandlerModule = (function () {
    function RequestHandlerModule() {
    }
    /**
     * @param {?} handlerConfig
     * @return {?}
     */
    RequestHandlerModule.forRoot = function (handlerConfig) {
        return {
            ngModule: RequestHandlerModule,
            providers: [
                {
                    provide: HANDLER_CONFIG,
                    useValue: handlerConfig
                },
                RequestHandlerService,
                {
                    provide: http.Http,
                    useFactory: InterceptorFactory,
                    deps: [
                        http.XHRBackend,
                        http.RequestOptions,
                        RequestHandlerService,
                    ],
                },
            ]
        };
    };
    /**
     * @param {?} handlerConfig
     * @return {?}
     */
    RequestHandlerModule.forTesting = function (handlerConfig) {
        return {
            ngModule: RequestHandlerModule,
            providers: [
                {
                    provide: HANDLER_CONFIG,
                    useValue: handlerConfig
                },
                RequestHandlerService,
                {
                    provide: http.Http,
                    useFactory: InterceptorFactory,
                    deps: [
                        http.XHRBackend,
                        http.RequestOptions,
                        RequestHandlerService,
                    ],
                },
                fakeBackendProvider,
                testing.MockBackend,
                http.BaseRequestOptions
            ]
        };
    };
    return RequestHandlerModule;
}());
RequestHandlerModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [
                    common.CommonModule,
                    http.HttpModule
                ],
                declarations: [],
                exports: []
            },] },
];
/**
 * @nocollapse
 */
RequestHandlerModule.ctorParameters = function () { return []; };

exports.InterceptorFactory = InterceptorFactory;
exports.RequestHandlerModule = RequestHandlerModule;
exports.RequestHandlerService = RequestHandlerService;

Object.defineProperty(exports, '__esModule', { value: true });

})));
